central/rbac/service/list_subjects.go

```diff
@@ -137,28 +140,35 @@ func NewSubjectSearcher(k8sRoleBindingDatastore datastore.DataStore) *SubjectSea
 
 // Search implements the searcher interface
 func (s *SubjectSearcher) Search(ctx context.Context, q *v1.Query) ([]search.Result, error) {
-	subjectQuery, _ := search.FilterQueryWithMap(q, mapping.OptionsMap)
-	pred, err := subjectFactory.GeneratePredicate(subjectQuery)
+	roleBindingQuery, _ := search.FilterQueryWithMap(q, schema.RoleBindingsSchema.OptionsMap)
+	roleBindingQuery, _ = search.InverseFilterQueryWithMap(roleBindingQuery, subjectMapping.OptionsMap)
+	roleBindingPred, err := roleBindingPredFactory.GeneratePredicate(roleBindingQuery)
 	if err != nil {
 		return nil, err
 	}
 
-	bindings, err := s.k8sRoleBindingDatastore.SearchRawRoleBindings(ctx, q)
+	subjectQuery, _ := search.FilterQueryWithMap(q, subjectMapping.OptionsMap)
+	subjectPred, err := subjectFactory.GeneratePredicate(subjectQuery)
 	if err != nil {
 		return nil, err
 	}
-	// Subject return only users and groups, there is a separate resolver for service accounts.
-	subjects := k8srbac.GetAllSubjects(bindings, storage.SubjectKind_USER, storage.SubjectKind_GROUP)
-	// Sort first then evaluate to not run evaluation twice.
-	// Sorting should be cheaper than reflect based evaluation
-	if err := sortSubjects(q, subjects); err != nil {
+
+	bindings, err := s.k8sRoleBindingDatastore.SearchRawRoleBindings(ctx, q)
```

Comment:
Pass `roleBindingQuery`

Reply:
roleBindingQuery has also applied inverse filter with subject options. So passing it would end up matching more bindings than expected.
`roleBindingQuery, _ = search.InverseFilterQueryWithMap(roleBindingQuery, subjectMapping.OptionsMap)`

The `InverseFIlter` is needed because without it, `roleBindingPred.Evaluate` will match subject values with keys `k8srolebinding.subjects.name` or `k8srolebinding.subjects.kind` . Removing those from search results while keeping `subject.kind` and `subject.name` that are matched by  `subjectPred.Evaluate` would lead to using too many string constants that depend on k8sRoleBinding and subject protos. The InverseFilter avoids this double matching out of the box without having to manually remove things from search results.